spanner/google/cloud/spanner/streamed.py

```diff
@@ -122,17 +130,49 @@ def _merge_values(self, values):
             field = self.fields[index]
             self._current_row.append(_parse_value_pb(value, field.type))
             if len(self._current_row) == width:
-                self._rows.append(self._current_row)
+                self._pending_rows.append(self._current_row)
                 self._current_row = []
 
+    def _flush_pending_rows(self):
+        """Helper for :meth:`consume_next`."""
+        flushed = self._pending_rows[:]
+        self._pending_rows[:] = ()
+        self._complete_rows.extend(flushed)
+
+    def _do_restart(self):
+        """Helper for :meth:`consume_next`."""
+        if not self._resume_token:
+            raise ValueError("No resume token")
+
+        self._pending_chunk = None
+        self._pending_rows[:] = ()
+        self._current_row[:] = ()
+        self._response_iterator = self._restart(self._resume_token)
+
     def consume_next(self):
```

Comment:
Can you add some unit tests for this class for the following cases:

1. Stream does not break.
2. Stream breaks before yielding any results.
3. Stream break after yielding a few results none of which had resume token.
4. Stream breaks when the last result had a resume token.
5. Stream breaks after a few rows have been yielded to the user and when the last result did not have a resume toekn

Any other cases you can think of.

Reply:
@vkedia 7d83816 updates the tests to assert that the `_do_restart` helper calls the `_restart` partial function without any arguments when the instance has no token yet.  The other cases are covered by existing unit tests.